Add netty bundle changes

isaacrivriv opened the pull request:
fixes #21349

wtlucy committed:
improve netty framework outbound connection support

wtlucy committed:
TLS support for netty endpoints

    pick 820985f160 add Netty support for sipcontainer

epj commented:
At the moment files picked for this pr don't match latest version in pr 20406 (Netty for Sip).   I'll post a few differences.  We can discuss what level of files makes best sense for this pr.

epj commented:
TCPUtils.java differences:

```12a13,15
> import java.net.Inet6Address;
> import java.net.InetSocketAddress;
> import java.net.SocketAddress;
31a35,36
> import io.openliberty.netty.internal.impl.NettyConstants;
>
35c40
<     private static final TraceComponent tc = Tr.register(NettyFrameworkImpl.class, TCPMessageConstants.NETTY_TRACE_NAME,
---
>     private static final TraceComponent tc = Tr.register(TCPUtils.class, TCPMessageConstants.NETTY_TRACE_NAME,
85a91,93
>         if (inetHost.equals("*")) {
>             inetHost = NettyConstants.INADDR_ANY;
>         }
94a103
>         final String newHost = inetHost;
99c108
<                 // add the new channel to the set of active channels, and set a close future to
---
>                 // add new channel to set of active channels, and set a close future to
107c116
<                 channel.attr(ConfigConstants.HostKey).set(inetHost);
---
>                 channel.attr(ConfigConstants.HostKey).set(newHost);
110a120,134
>                 // set up a helpful log message
>                 String hostLogString = newHost == NettyConstants.INADDR_ANY ? "*" : newHost;
>                 SocketAddress addr = channel.localAddress();
>                 InetSocketAddress inetAddr = (InetSocketAddress)addr;
>                 String IPvType = "IPv4";
>                 if (inetAddr.getAddress() instanceof Inet6Address) {
>                     IPvType = "IPv6";
>                 }
>                 if (newHost == NettyConstants.INADDR_ANY) {
>                     hostLogString = "*  (" + IPvType + ")";
>                 } else {
>                     hostLogString = config.getHostname() + "  (" + IPvType + ": "
>                                + inetAddr.getAddress().getHostAddress() + ")";
>                 }
>
113c137
<                             new Object[] { config.getExternalName(), inetHost, String.valueOf(inetPort) });
---
>                             new Object[] { config.getExternalName(), hostLogString, String.valueOf(inetPort) });
117c141
<                                 new Object[] { config.getExternalName(), inetHost, String.valueOf(inetPort) });
---
>                                 new Object[] { config.getExternalName(), hostLogString, String.valueOf(inetPort) });
140c164
<                     open(framework, bootstrap, config, inetHost, inetPort, openListener, retryCount - 1);
---
>                     open(framework, bootstrap, config, newHost, inetPort, openListener, retryCount - 1);
143c167
<                         Tr.error(tc, TCPMessageConstants.BIND_ERROR, new Object[] { config.getExternalName(), inetHost,
---
>                         Tr.error(tc, TCPMessageConstants.BIND_ERROR, new Object[] { config.getExternalName(), newHost,
148c172
<                                     inetHost, String.valueOf(inetPort), openFuture.cause().getMessage() });
---
>                                     newHost, String.valueOf(inetPort), openFuture.cause().getMessage() });```

epj commented:
UDPUtils.java differences:

```12a13
> import java.net.Inet6Address;
13a15
> import java.net.SocketAddress;
27a30
> import io.openliberty.netty.internal.impl.NettyConstants;
31c34
<     private static final TraceComponent tc = Tr.register(NettyFrameworkImpl.class, UDPMessageConstants.NETTY_TRACE_NAME,
---
>     private static final TraceComponent tc = Tr.register(UDPUtils.class, UDPMessageConstants.NETTY_TRACE_NAME,
70a74,78
>         if (inetHost.equals("*")) {
>             inetHost = NettyConstants.INADDR_ANY;
>         }
>         final String newHost = inetHost;
>
88c96
<                 channel.attr(ConfigConstants.HostKey).set(inetHost);
---
>                 channel.attr(ConfigConstants.HostKey).set(newHost);
91a100,115
>                 // set up a helpful log message
>                 String hostLogString = newHost == NettyConstants.INADDR_ANY ? "*" : newHost;
>                 SocketAddress addr = channel.localAddress();
>                 InetSocketAddress inetAddr = (InetSocketAddress)addr;
>                 String IPvType = "IPv4";
>                 if (inetAddr.getAddress() instanceof Inet6Address) {
>                     IPvType = "IPv6";
>                 }
>                 if (newHost == NettyConstants.INADDR_ANY) {
>                     hostLogString = "*  (" + IPvType + ")";
>                 } else {
>                     hostLogString = config.getHostname() + "  (" + IPvType + ": "
>                                + inetAddr.getAddress().getHostAddress() + ")";
>                 }
>
>
95c119
<                             new Object[] { config.getExternalName(), inetHost, String.valueOf(inetPort) });
---
>                             new Object[] { config.getExternalName(), hostLogString, String.valueOf(inetPort) });
99c123
<                                 new Object[] { config.getExternalName(), inetHost, String.valueOf(inetPort) });
---
>                                 new Object[] { config.getExternalName(), hostLogString, String.valueOf(inetPort) });
121c145
<                     bind(framework, bootstrap, inetHost, inetPort, retryCount - 1, retryDelay, bindListener);
---
>                     bind(framework, bootstrap, newHost, inetPort, retryCount - 1, retryDelay, bindListener);
125c149
<                                 new Object[] { channelName, inetHost, String.valueOf(inetPort) });
---
>                                 new Object[] { channelName, newHost, String.valueOf(inetPort) });
129c153
<                                     new Object[] { channelName, inetHost, String.valueOf(inetPort) });
---
>                                     new Object[] { channelName, newHost, String.valueOf(inetPort) });
188c212
<                         newHost = "0.0.0.0";
---
>                         newHost = NettyConstants.INADDR_ANY;
189a214
>                     String hostLogString = newHost == NettyConstants.INADDR_ANY ? "*" : newHost;
195c220
<                                 new Object[] { channelName, newHost, String.valueOf(inetPort) });
---
>                                 new Object[] { channelName, hostLogString, String.valueOf(inetPort) });
197c222
<                                 "local address unresolved for " + channelName + " - " + newHost + ":" + inetPort);
---
>                                 "local address unresolved for " + channelName + " - " + hostLogString + ":" + inetPort);
```

epj commented:
UDPConfigurationImpl.java differences:

```23a24
> import io.netty.channel.FixedRecvByteBufAllocator;
88a90
>             //SO_RCVBUF - size of buffer that holds the datagrams the client hasn't read yet
89a92,93
>             //set common RCVBUF_ALLOCATOR strategy
>             bootstrap.option(ChannelOption.RCVBUF_ALLOCATOR, new FixedRecvByteBufAllocator(getReceiveBufferSize()));
```

isaacrivriv committed:
Add outbound support and additional changes to Netty bundle

isaacrivriv committed:
Fix alias Netty Outbound SSL and added Netty log TCP channel started

isaacrivriv commented:
#build

mrsaldana commented:
First impression is that this is fine, can be built on top of and doesn't affect any existing externalized feature. I think there are a few places we will eventually revisit, like ChannelOptions / Buffer policy, but I don't have any strong resistance as to not merging this since it is just updating the already existing Netty bundle with the more recent changes

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -161,7 +147,7 @@ public void serverStopping() {
             for (Channel channel : activeChannels) {
                 channel.closeFuture().addListener(future -> {
-                    logChannelStopped(channel);
+                    stop(channel);
```
logChannelStopped is still used elsewhere, is it too early here or why?

isaacrivriv replied:
I believe in this case this was a mistake that was corrected since calling only logChannelStopped doesn't actually stop the channel only logs it which means we could still have references to the channel still going. By calling stop all the references we keep are ensured to be cleared and it eventually also calls logChannelStopped so we should be good to go with the correction

hutchig replied:
OK understood!

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -304,37 +300,53 @@ public ChannelFuture start(BootstrapExtended bootstrap, String inetHost, int ine
-                        Tr.debug(tc,
-                                "channel " + channelName + " failed to stop due to: " + future.cause().getMessage());
+        if (activeChannels.contains(channel)) {
+            activeChannels.remove(channel);
```
swaps the order of close and remove from active compared to old code?

hutchig replied:
also now if serverStopping has been called which set isActive to false a call to this method
will remove the channel from the active set but will not close it.

isaacrivriv replied:
Yeah this seems a bit weird I'll need to revisit this and see what the logic was to change this order

isaacrivriv replied:
I've gone through it but I'm not sure if this actually messes anything up due to the fact that activeChannels is thread safe but there could be a pathway where this is possible. I created https://github.com/OpenLiberty/open-liberty/issues/23441 to follow up on this and investigate further.

hutchig replied:
OK (I think - worth talking through and raising a followup issue is fair enough.)

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -304,37 +300,53 @@ public ChannelFuture start(BootstrapExtended bootstrap, String inetHost, int ine
     @Override
     public void stop(Channel channel, long timeout) {
-        if (this.isActive && channel.isOpen()) {
+        if (activeChannels.contains(channel)) {
```
There is a race here perhaps if activeChannels membership is used as the gate - is it atomic etc?

isaacrivriv replied:
activeChannels is a ConcurrentHashMap which is thread safe so think we should be covered

hutchig replied:
OK

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,276 @@
+        } else {
+            oFuture = ((BootstrapExtended) bootstrap).connect(inetHost, inetPort);
+        }
+        final ChannelFuture openFuture = oFuture;
```
why oFuture and openFuture too?

isaacrivriv replied:
Seems a bit redundant, can't think of a reason why this was done this way. I'll look more into it and make the change if nothing gets affected

hutchig replied:
OK

hutchig replied:
Having both of these is not necessary

isaacrivriv replied:
Following up on this [here](https://github.com/OpenLiberty/open-liberty/issues/24840)

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,276 @@
+                channel.attr(ConfigConstants.IsInboundKey).set(config.isInbound());
+
+                // set up a helpful log message
+                String hostLogString = newHost == NettyConstants.INADDR_ANY ? "*" : newHost;
```
This line has no effect due to further assignment below

isaacrivriv replied:
Nice catch, I'll fix this one too

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,254 @@
+                channel.attr(ConfigConstants.IsInboundKey).set(config.isInboundChannel());
+                
+                // set up a helpful log message
+                String hostLogString = newHost == NettyConstants.INADDR_ANY ? "*" : newHost;
```
value not used as reassigned below

isaacrivriv replied:
Nice catch, I'll fix this

hutchig commented on dev/io.openliberty.netty.internal.impl/test/io/openliberty/netty/internal/impl/NettyFrameworkImplTest.java:
```diff
@@ -167,6 +168,91 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf buf) throws Excep
+        setTCPConfig();
+        ServerBootstrapExtended bootstrap = framework.createTCPBootstrap(options);
+        bootstrap.childHandler(bootstrap.getBaseInitializer());
+        final CountDownLatch latch = new CountDownLatch(1);
```
I always find variables named like this a bit difficult to pass by on a review, here it does not really matter as it is used so nearby but I think it is a good habit to give it a descriptive name. Up to you of course as not a defect.

isaacrivriv replied:
I took a look and I agree it could be hard to follow up on but since this is just a unit test, I decided to leave as is for the time being. We can revisit this more in the future and make the updates to make this better to understand.

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,327 @@
+
+        SSLContext jdkContext;
+        try {
+        	// TODO: Check alias being sent here
```
Can't merge in TODO's ? Raise an issue etc?

isaacrivriv replied:
Yep you're right nice catch, I'll work on clearing this TODO up

isaacrivriv replied:
Issue created to follow up https://github.com/OpenLiberty/open-liberty/issues/23424

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,327 @@
+                    }
+                });
+            } catch (Exception e) {
+                // no ffdc required
```
Seems unusual to swallow such a wide exception?

isaacrivriv replied:
It does seem a bit unusual. The TLS implementation was taken and adapted from the original channel code in SSLChannel [here](https://github.com/OpenLiberty/open-liberty/blob/2fd4a880754c37a988c5ed9ac4f1ea5988e465d6/dev/com.ibm.ws.channel.ssl/src/com/ibm/ws/channel/ssl/internal/SSLChannel.java#L372) from what I can deduce in the code I believe it should be fine and that it is expected in some occasions

hutchig replied:
OK

isaacrivriv committed:
Worked on other changes to bundle
- Fixed comments from review and cleaned up
- Added logger to match channelfw TCP logging connection lifecycle

mrsaldana commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/ConfigConstants.java:
```diff
@@ -14,9 +14,11 @@
-    public AttributeKey<String> HostKey = AttributeKey.valueOf("Host");
-    public AttributeKey<String> ExternalHostKey = AttributeKey.valueOf("ExternalHost");
-    public AttributeKey<Integer> PortKey = AttributeKey.valueOf("Port");
+    public final String ExternalName = "ExternalName";
```
Consider String Constant convention EXTERNAL_NAME

isaacrivriv replied:
Sounds good with me, I'll make this change

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/LibertyLoggingHandler.java:
```diff
@@ -0,0 +1,161 @@
+	protected static final TraceComponent tc = Tr.register(LibertyLoggingHandler.class, new String[]{TCPMessageConstants.TCP_TRACE_NAME,TCPMessageConstants.NETTY_TRACE_NAME},
+			TCPMessageConstants.TCP_BUNDLE, LibertyLoggingHandler.class.getName());
+	
+	protected final static AttributeKey<Integer> BYTES_READ_KEY = AttributeKey.valueOf("BYTES_READ");
```
Do you foresee this being used by any other class? It might be worth considering moving this definition to a place where it can be called into more generally within the netty internal package

isaacrivriv replied:
Don't think I see this being used anywhere else right now besides the TCP pipeline. The name of the class is misleading, I'll change to something more related to TCP to address this

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/LibertyLoggingHandler.java:
```diff
@@ -0,0 +1,161 @@
+	@Override
+	public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+			Tr.event(ctx.channel(), tc, "registered " + ctx.channel() + ", key is " + null); //TODO Can't get selection key for Netty, check if this is actually necessary
```
If key is not obtainable at this point, it is likely better to not include it. Having it say 'key is null' might be confusing from an L2 perspective.

isaacrivriv replied:
I agree, I'll make the change to remove this

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/LibertyLoggingHandler.java:
```diff
@@ -0,0 +1,161 @@
+			ctx.channel().attr(BYTES_READ_KEY).getAndSet(ctx.channel().attr(BYTES_READ_KEY).get() + ((ByteBuf) msg).readableBytes());
+		}else {
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(ctx.channel(), tc, "Could not get bytes for object: " + msg);
```
Are there any  plans to support non ByteBuf object reads/write? Should we consider this an error case/unsupported scenario? I understand that the position of the handler in the pipeline determines the typing that can be passed on, but I'd imagine that we would aim to position this handler in a consistent fashion and passing along the ByteBuf used to read/write data.

isaacrivriv replied:
I'll adapt the class access modifier so that it should only be used inside the package itself. This should help ensure that we don't hit this issue so I'll also adapt the code to log as if we only expect ByteBuf objects. Thinking about this in a more broader sense, we should discuss ways to ensure that the pipeline is not modified outside the bundle which could affect things like timeouts and logging. Created https://github.com/OpenLiberty/open-liberty/issues/23435 to follow up on this

isaacrivriv committed:
Fixed additional comments
- Made changes to constants following upper case snake case
- Updated logging handler to be TCP specific

isaacrivriv commented:
#build

hutchig reviewed (dismissed):
Thank for looking at the code from the review comments. I can't see anything else
to comment on so - LGTM. If I EVER do not respond to a review for more than 1 working
day (and you don't have the review you need from anyone else) then feel free to DM me
as it will not be intentional that I do this silently.

isaacrivriv committed:
Rebased and updated copyright date and version if needed to EPL 2.0

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPConfigurationImpl.java:
```diff
@@ -372,7 +377,8 @@ public void setChannelReceiveBufferSize(int size) {
         if (size < 0 || size > UDPConfigConstants.MAX_UDP_PACKET_SIZE) {
             if (tc.isDebugEnabled()) {
-                Tr.debug(tc, "Channel Receive buffer size not within Limits: " + size + " setting to default: " + UDPConfigConstants.MAX_UDP_PACKET_SIZE);
+                Tr.debug(tc, "Channel Receive buffer size not within Limits: " + size + " setting to default: "
```
I would have said 'setting to maximum' would be better.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPConfigurationImpl.java:
```diff
@@ -83,13 +85,17 @@ public void applyConfiguration(Bootstrap bootstrap) {
         bootstrap.option(ChannelOption.SO_REUSEADDR, true);
 
         if ((getReceiveBufferSize() >= UDPConfigConstants.RECEIVE_BUFFER_SIZE_MIN)
```
If getReceiveBufferSize() is cheap then calling it 4 times in a blink of an eye might not add much time but, when writing middleware, if it was in the send receive path and not in bootstrap, and even though the JIT etc might be able to inline it, it is not a great look.

hutchig replied:
5 times (missed one).

isaacrivriv replied:
Updated

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+     * @return
+     * @throws NettyException
+     */
+    public static BootstrapExtended createUDPBootstrap(NettyFrameworkImpl framework, Map<String, Object> options)
```
Perhaps the method would be better called ...Inbound

isaacrivriv replied:
See comment below https://github.com/OpenLiberty/open-liberty/pull/21350#discussion_r1149628843

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+        return create(framework, new UDPConfigurationImpl(options, false));
+    }
+
+    private static BootstrapExtended create(NettyFrameworkImpl framework, UDPConfigurationImpl config) {
```
If this is not creating a UDPUtils facade (ie a creation factory for the class it is in) then "create what"? So it would be better called createBoostrapExtended. Just an opinion, as are all my comments so far so don't feel you have to change it. Also, I know the 'external' public methods do have meaningful names already, but as a 'Utils' class this class isn't going to be dealing with just one sort of 'thing' internally perhaps.

isaacrivriv replied:
See comment below https://github.com/OpenLiberty/open-liberty/pull/21350#discussion_r1149628843

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+
+    private static ChannelFuture bind(NettyFrameworkImpl framework, BootstrapExtended bootstrap, String inetHost,
+            int inetPort, final int retryCount, final int retryDelay, ChannelFutureListener bindListener) {
+        ChannelFuture bindFuture = bootstrap.bind(inetHost, inetPort);
```
Interesting that we use inetHost 'as is' * passing it into bind() and **then** we set it to INADDR_ANY.
So methods have to be able to deal with "*" and/or INADDR_ANY or just some or just bind()?
It does not need to change, but sometimes it is better to use one of two alternatives everywhere,
getting rid of the either/or as soon as possible and being consistent afterwards.

isaacrivriv replied:
We may want to follow up on this on https://github.com/OpenLiberty/open-liberty/issues/18520

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+                // add the new channel to the set of active channels, and set a close future to
+                // remove it
+                final Channel channel = bindFuture.channel();
+                framework.getActiveChannels().add(bindFuture.channel());
```
it would be better to use the variable you just got?

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+                channel.closeFuture().addListener(innerFuture -> framework.stop(channel));
+
+                // set common channel attrs
+                channel.attr(ConfigConstants.NAME_KEY).set(config.getExternalName());
```
You should set the state of channel BEFORE setting up a listener, 99.999% of time (e.g. during testing) the stop method with be able to use/trace the name, host and port and then one day it won't.

isaacrivriv replied:
Understood, updated to add the listener after setting common attributes

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,256 @@
+     * @return
+     * @throws NettyException
+     */
+    public static ChannelFuture start(NettyFrameworkImpl framework, BootstrapExtended bootstrap, String inetHost,
```
Would it be better startInbound?

isaacrivriv replied:
There's a couple of references that can definitely be clarified/updated. I would like to do this work while working on updating the SIP Netty changes so created https://github.com/OpenLiberty/open-liberty/issues/24840 to follow up on this

hutchig commented on dev/io.openliberty.netty.internal.impl/test/io/openliberty/netty/internal/impl/NettyFrameworkImplTest.java:
```diff
@@ -169,6 +170,91 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf buf) throws Excep
+        });
+        framework.setServerStarted(null);
+        latch.await(PORT_BIND_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        Thread.sleep(10); // wait to ensure framework's internal callbacks are handled
```
If this is run as the unit tests in a build you might want to add a good slug into 10ms to cope with fyre sluggishness?

isaacrivriv replied:
Nice idea, updated to 1000ms

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+    static final String ALIAS_KEY = "alias";
+    
+    /**
+     * DS deactivate
```
but it is the activate method :-(

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+    }
+    
+    /**
+     * DS deactivate
```
matches method name?

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+            }
+        } catch (Exception e) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "getOutboundSSLContext exception caught creating SSLContext: " + e);
```
I think this should be traced higher than debug and at a level that would appear in a standard customer trace file that would come in if the error occurred?

hutchig replied:
Surely this failing is a 'warning'?

isaacrivriv replied:
Updated to be a warning message

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+            return nettyContext;
+        } catch (Exception e) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "getOutboundSSLContext exception caught creating JdkSslContext: " + e);
```
I think this should be traced higher than debug and at a level that would appear in a standard customer trace file that would come in if the error occurred

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    /**
+     * Build a {@link io.netty.handler.ssl.SslContext} for an intbound connection
```
intbound

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+            return nettyContext;
+        } catch (Exception e) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "getInboundSSLContext exception caught creating JdkSslContext: " + e);
```
This should be higher than 'debug' as it is an unexpected problem, warning?/error?

isaacrivriv replied:
I agree, I'll update this to be a warning

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+            } catch (
+
+            Exception e) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
```
A general comment on having isDebugEnabled() and tr.debug when catching unexpected exceptions that will stop the server working as customer expects. Errors/Warnings, customer turns trace on?

isaacrivriv replied:
Similar to the next comment, this is based off the SSL creation code in com.ibm.ws.channel.ssl [here](https://github.com/OpenLiberty/open-liberty/blob/e09e937bd8e87ee850c82fc5b6f8a50a736edb28/dev/com.ibm.ws.channel.ssl/src/com/ibm/ws/channel/ssl/internal/SSLChannel.java#L483) used also by channelfw. I created https://github.com/OpenLiberty/open-liberty/issues/24838 to follow up on this a bit more

hutchig replied:
ok

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,326 @@
+        } catch (Exception e) {
+            // no FFDC required
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "Exception getting SSLContext from properties.", new Object[] { e });
```
should not be jsut debug and should include some of the key variables?

isaacrivriv replied:
I agree it could probably use some additional details. This is based off the SSL creation code in com.ibm.ws.channel.ssl [here](https://github.com/OpenLiberty/open-liberty/blob/e09e937bd8e87ee850c82fc5b6f8a50a736edb28/dev/com.ibm.ws.channel.ssl/src/com/ibm/ws/channel/ssl/internal/SSLChannel.java#L483) used also by channelfw. Not sure how closely related we want both of these to be so I created https://github.com/OpenLiberty/open-liberty/issues/24838 to follow up on this a bit more

hutchig replied:
ok

hutchig commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/ConfigConstants.java:
```diff
@@ -16,9 +16,11 @@
-    public AttributeKey<String> ExternalHostKey = AttributeKey.valueOf("ExternalHost");
-    public AttributeKey<Integer> PortKey = AttributeKey.valueOf("Port");
+    public final String EXTERNAL_NAME = "ExternalName";
+    public final AttributeKey<String> NAME_KEY = AttributeKey.valueOf(EXTERNAL_NAME);
```
Looks odd just having one of these strings and not having EXTERNAL_NAME_KEY but NAME_KEY  when there is EXTERNAL_HOST_KEY ?

isaacrivriv replied:
It looks odd indeed, I believe it was because it is used in another piece of code. I'm updating to add string keys for all the other attributes as well to make it consistent.

hutchig replied:
ok

hutchig reviewed (dismissed):
Please read comments.

isaacrivriv committed:
Fixed issues with more than one netty feature removing the public netty feature and addressed comments from PR

isaacrivriv commented:
#build

pnicolucci commented on dev/com.ibm.websphere.appserver.features/visibility/private/io.openliberty.netty-1.0.feature:
```diff
@@ -1,18 +1,15 @@
 -include= ~${workspace}/cnf/resources/bnd/feature.props
 symbolicName=io.openliberty.netty-1.0
```
Since we're moving this to a private feature we need to update the feature name to follow the open liberty naming guidelines: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-naming-guidelines

isaacrivriv replied:
I'll update to be io.openliberty.netty.internal-1.0

pnicolucci commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/tls/package-info.java:
```diff
@@ -0,0 +1,12 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
```
This new line in the copyright header isn't needed.

isaacrivriv replied:
I'll update all of these

pnicolucci commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/tls/NettyTlsProvider.java:
```diff
@@ -0,0 +1,21 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
```
This new line in the copyright header isn't needed.

pnicolucci commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/ChannelInitializerWrapper.java:
```diff
@@ -1,3 +1,13 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
```
This new line in the copyright header isn't needed.

pnicolucci commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/package-info.java:
```diff
@@ -0,0 +1,12 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
```
This new line in the copyright header isn't needed.

pnicolucci commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,285 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
```
This new line in the copyright header isn't needed.

pnicolucci commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPLoggingHandler.java:
```diff
@@ -0,0 +1,156 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * 
```
This new line in the copyright header isn't needed.

isaacrivriv committed:
Fixed review comment
- Updated to add names for handlers
- Fixed comments
- Added fix for inactivity handler

isaacrivriv commented:
#build

isaacrivriv commented:
#build

epj commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPMessageConstants.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
There are a few files with 2022.   Any being updated by this pr should be dated 2023.

isaacrivriv replied:
We should probably discuss this a bit more. Last discussion we had about this I planned on updating everything to the latest year (2023) but since the files/changes were introduced beforehand and not updated (only rebased commits, etc) then we decided to only update the ones that actually had changes.

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPMessageConstants.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,284 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.tls.impl/bnd.bnd:
```diff
@@ -0,0 +1,36 @@
+#*******************************************************************************
+# Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal/bnd.bnd:
```diff
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2021 IBM Corporation and others.
+# Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/NettyFramework.java:
```diff
@@ -1,90 +1,135 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/tls/NettyTlsProvider.java:
```diff
@@ -0,0 +1,20 @@
+/*******************************************************************************
+ * Copyright (c) 2021 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal/src/io/openliberty/netty/internal/tls/package-info.java:
```diff
@@ -0,0 +1,11 @@
+/*******************************************************************************
+ * Copyright (c) 2021 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/package-info.java:
```diff
@@ -0,0 +1,11 @@
+/*******************************************************************************
+ * Copyright (c) 2021 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPLoggingHandler.java:
```diff
@@ -0,0 +1,155 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPConfigurationImpl.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
Copyright year

mrsaldana commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/MaxOpenConnectionsHandler.java:
```diff
@@ -1,14 +1,11 @@
 /*******************************************************************************
- * Copyright (c) 2021 IBM Corporation and others.
+ * Copyright (c) 2021, 2022 IBM Corporation and others.
```
copyright year

hutchig commented on dev/io.openliberty.netty.internal.impl/bnd.bnd:
```diff
@@ -25,6 +25,7 @@ Export-Package: \
   io.openliberty.netty.internal.udp
 
 Import-Package: \
+  com.ibm.ws.channelfw.internal.chains, \
```
Might be worth raising an issue to handle Endpoint manager implementation teasing out of cfw longer term. Interesting that there is an io.openliberty.endpoint project separate from cfw - wonder if that is part of the solution?

hutchig replied:
I think, as there is already a separate bundle - this is just the package name.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java:
```diff
@@ -359,8 +368,16 @@ public void destroy() {
+
     private void logChannelStopped(Channel channel) {
-        if (channel instanceof NioServerSocketChannel) {
+        if (channel instanceof NioServerSocketChannel || channel instanceof NioSocketChannel) {
```
Longer term it would probably be better to have different UDP  or TCP based channels to
have a marker interface so all the subtypes did not have to be in if statements here (as it shoulod not have to know all te subtypes), or better still
have a 'loggable' abstract interface that all channels have to implement themselves with a
logStopped method etc. and call that on the channel to then delegate off to UDP/TCPUtils etc. as IT sees fit.

isaacrivriv replied:
I think with the latest changes `logChannelStopped` isn't used anymore. I'll follow up on this to clean up what we can [here](https://github.com/OpenLiberty/open-liberty/issues/24840)

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPLoggingHandler.java:
```diff
@@ -0,0 +1,155 @@
+			public void operationComplete(ChannelFuture future) throws Exception {
+				if(future.isSuccess())
+					Tr.event(ctx.channel(), tc, "write (async) finished sucessfully for local: " + ctx.channel().localAddress() + " remote: " + ctx.channel().remoteAddress()+ " wrote: " + bytesToWrite + " bytes");
+				else
```
the value of future isCancelled and cause() should be logged (or the fact it is null if so)

hutchig reviewed (dismissed):
Approved with some minor comments.

isaacrivriv commented:
#build

isaacrivriv commented:
#build

isaacrivriv committed:
Added additional Netty functionality
- Updated/Fixed API to allow for Future tasks instead of channel futures
- Added quiesce functionality for Netty channels with channel tracking
- Changed groups of connections for UDP
- Updated tests to run with changes

isaacrivriv commented:
#build

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java:
```diff
@@ -0,0 +1,138 @@
+	            	Tr.debug(tc, "Remaining endpoints still running. Verifying which can be closed...");
+	            }
+				// Iterate through active endpoints
+				for(Channel endpoint : activeChannels) {
```
I am not sure "endpoint" is a great name for this as an Endpoint is something else.

hutchig replied:
'ch' 'channel' 'activeChannel' perhaps?

isaacrivriv replied:
In this case `endpoint` pretty much refers to the same meaning as the already used Endpoint. The already used Endpoint is just the descriptor while the new endpoint variable here points to the actual channel for an Endpoint. They are different objects in classes but refer to the same concept. I couldn't think of anything else that would describe this in a better way so I would like to keep as is unless there's another term that could be used.

isaacrivriv replied:
Perhaps activeChannels is something that we could change to activeEndpoints maybe that could simplify things since activeChannel no longer contain all the channels but the actual channel endpoints

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java:
```diff
@@ -0,0 +1,138 @@
+					// Verify if there are still active connections for that endpoint
+					ChannelGroup group = this.bundle.getActiveChannelsMap().get(endpoint);
+					if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+		            	Tr.debug(tc, "Found group " + group + " for endpoint " + endpoint);
```
"for channel" ?

isaacrivriv replied:
Same thing here, we only quiesce the endpoint channels to follow the same CHFW quiesce behavior. Perhaps renaming the getActiveChannelsMap to something like getActiveEndpointsMap could be more useful

hutchig replied:
OK. Given the other comments about endpoint/channels it seems self-consistent so fine as is.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java:
```diff
@@ -0,0 +1,138 @@
+	
+	private ScheduledExecutorService scheduler;
+	
+	private ScheduledFuture<?> quiesceFinishTask;
```
I think that using the same postfix "Task" for both Runnables and Futures is confusing - it would be better as ...Future maybe?

isaacrivriv replied:
Not sure I understand this one. Can you clarify?

hutchig replied:
What I was trying to say was that this variable is not actually a Task (ie. a runnable/callable thing) it is the Future that relates to the task's result, so calling it a name that ends in Task is inaccurate. If this was good then MOST future variable names could end in 'Task' (and still none of them would be Tasks - they would still all be Futures.) It is not a hard problem though - so if you don't agree, I will not say any more (that is why I selected Approve on the review.)

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java:
```diff
@@ -0,0 +1,138 @@
+		
+		this.quiesceFinishTask = scheduler.schedule( ()->{
+			if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            	Tr.debug(tc, "The quiesceFinishTask has started for " + this.bundle + ". Quiesce task will be scheduled to run for: " + this.bundle);
```
(it is not the Future that has started but the Runnable)

isaacrivriv replied:
Something that we need to change in the log message?

hutchig replied:
I think I was whining about the name of the variable not differentiating the Task and the Future again - safe to not change the log msg.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyQuiesceListener.java:
```diff
@@ -0,0 +1,138 @@
+            	Tr.debug(tc, "The quiesceVerifierTask has started,  checking endpoints...");
+            }
+			Set<Channel> activeChannels = this.bundle.getActiveChannels();
+			if(activeChannels.isEmpty()) {
```
Channels or Connections? Better to use one term for the same thing.

isaacrivriv replied:
In this case, connections refer to the active connections to the endpoints (child channels started from the main server channel). Since channels refer to pretty much any connection on Netty, I used this terminology to distinguish them: `Connections` for channels created from a ServerChannel or outbound connections, and `Endpoints` for TCP channels which are bind to a port. How can I make this more understandable? Maybe adding some comments to make sure they don't get confused?

hutchig replied:
I will happily adopt the same terminology if/as it makes sense, it can be good to inoculate such conventions with comments but I will catch up either way and try to be consistent.

epj reviewed (commented):
Not sure this is 'late cycle' here?  You also have this same message in NettyFrameworkImpl.

epj commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+                });
+            } catch (Exception e) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc, "caught exception performing late cycle server startup task: " + e.getMessage());
```
I was trying to point at this line with my previous comment... is this correct to say 'late cycle' here?

isaacrivriv replied:
I think it's referring to having issues starting up a task after server startup. CHFW does something similar [here](https://github.com/OpenLiberty/open-liberty/blob/12f6257b2c5820a289e1afff3f4e5e16fbc1aada/dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPPort.java#L315-L317C47) but it's not the same message. This could be change if we want to, maybe `NettyFramework signaled- caught exception::` could be better to follow similar logging as CHFW

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/QuiesceHandler.java:
```diff
@@ -0,0 +1,61 @@
+
+
+	public QuiesceHandler(Callable quiesceTask) {
+		this.quiesceTask = quiesceTask;
```
Please don;t feel the need to change this, but I would have called it  serverQuiesceCallable
(some '...tasks' elsewhere are Runnables)

isaacrivriv replied:
Understood, if I have to make any additional changes I'll switch up the name.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/MaxOpenConnectionsHandler.java:
```diff
@@ -49,7 +46,7 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
             long currentTime = System.currentTimeMillis();
             if (currentTime > (lastConnExceededTime + 600000L)) {
-                String channelName = ctx.channel().attr(ConfigConstants.NameKey).get();
+                String channelName = ctx.channel().attr(ConfigConstants.NAME_KEY).get();
```
wondered if we should use/implement toString of channel's class.

isaacrivriv replied:
The Netty Channel already has a overriden toString which concats the host port of local and remote but we use these names to mimic the CHFW behavior related to logging. We use attributes to append to a Channel in this case instead of an extension to the class mainly cause it's a cleaner/easier solution IMO

hutchig replied:
OK

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/MaxOpenConnectionsHandler.java:
```diff
@@ -49,7 +46,7 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
-                String channelName = ctx.channel().attr(ConfigConstants.NameKey).get();
+                String channelName = ctx.channel().attr(ConfigConstants.NAME_KEY).get();
                 Tr.warning(tc, TCPMessageConstants.MAX_CONNS_EXCEEDED, channelName, maxConnections);
                 lastConnExceededTime = currentTime;
```
If this is affecting externally visible behaviour - do we have a message in message.log or do people just 'silently' see connections not happening?

isaacrivriv replied:
There's the warning message we print MAX_CONNS_EXCEEDED which users should see. Maybe we need to look into this message being printed every time for every max connection happening, we would need to check the CHFW behavior to mimic it.

hutchig replied:
If there is a follow up issue number I will resolve this. I agree, we perhaps do not want many occurrences of the same message etc.

isaacrivriv replied:
I realized that CHFW does the same [thing](https://github.com/OpenLiberty/open-liberty/blob/fb0a620ce07a7eff6bea0aeaefe973b9d31ee337/dev/com.ibm.ws.channelfw/src/com/ibm/ws/tcpchannel/internal/TCPChannel.java#L626-L628), it notifies every 10 minutes so behavior stays the same. If you still think we need to look more into this, lets create an issue to follow up.

hutchig replied:
No that is fine, I will set as resolved.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPChannelInitializerImpl.java:
```diff
@@ -26,24 +29,57 @@
+    	// TODO Add logging equal to channelfw
+    	if(bundle.isStopping()) {
+    		if(TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+            	Tr.event(tc, "Tried to start channel: " + channel + " while framework was shutting down. " + bundle);
```
'bundle is stopping' is better here than 'framework is shutting down'

hutchig replied:
The act of shutting down is a lifecycle event but that a channel start clashed with it is more debug imho.
    int TRACE_LEVEL_DUMP = 0;
    int TRACE_LEVEL_DEBUG = 1;
    int TRACE_LEVEL_ENTRY_EXIT = 2;
    int TRACE_LEVEL_EVENT = 3;
    int TRACE_LEVEL_DETAIL = 4;
    int TRACE_LEVEL_CONFIG = 5;
    int TRACE_LEVEL_INFO = 6;
    int TRACE_LEVEL_AUDIT = 7;
    int TRACE_LEVEL_WARNING = 8;
    int TRACE_LEVEL_ERROR = 9;
    int TRACE_LEVEL_FATAL = 10;
    int TRACE_LEVEL_OFF = 11;

hutchig replied:
The act of shutting down is a lifecycle event but that a channel start clashed with it is more debug imho.
    int TRACE_LEVEL_DUMP = 0;
    int TRACE_LEVEL_DEBUG = 1;
    int TRACE_LEVEL_ENTRY_EXIT = 2;
    int TRACE_LEVEL_EVENT = 3;
    int TRACE_LEVEL_DETAIL = 4;
    int TRACE_LEVEL_CONFIG = 5;
    int TRACE_LEVEL_INFO = 6;
    int TRACE_LEVEL_AUDIT = 7;
    int TRACE_LEVEL_WARNING = 8;
    int TRACE_LEVEL_ERROR = 9;
    int TRACE_LEVEL_FATAL = 10;
    int TRACE_LEVEL_OFF = 11;

isaacrivriv replied:
This is a fair point and something we should look more into. From what I could see glancing in the CHFW bundle didn't see any logic where if the server is shutting down then new connections won't be started.

hutchig replied:
This is not important enough to warrant its own follow up issue IMHO and is current behaviour so resolving.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPChannelInitializerImpl.java:
```diff
@@ -26,24 +29,57 @@
+    		return;
+    	}
+        if (TraceComponent.isAnyTracingEnabled()) {
+        	channel.pipeline().addFirst(NettyConstants.TCP_LOGGING_HANDLER_NAME, new TCPLoggingHandler());
```
I would have not called the pipleline method so many times but used a pipeline variable

isaacrivriv replied:
I don't think it really matters considering this is a straight getter from the channel and from multiple Netty sources/examples they follow the same approach to get a channel pipeline.

hutchig replied:
I think you want to code for future subclassing etc generally and not assume repeated method calls are cheap, as a general style it looks 'less good' sometimes. At runtime the JIT is likely to remove them anyway and, you are as usual correct, and I take your point :-) so resolving

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPChannelInitializerImpl.java:
```diff
@@ -26,24 +29,57 @@
+        	ChannelGroup group = bundle.getActiveChannelsMap().get(parent);
+        	channel.closeFuture().addListener(innerFuture -> TCPUtils.logChannelStopped(channel));
+        	if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            	Tr.debug(tc, "Found group to be: " + group + " for parent: " + parent);
```
We are relying on the toString of parent(channel) here but in the MaxOpenConnections handler we reached inside for the name.
Msg might be better as something like 'adding channel C to group g of parent p'

isaacrivriv replied:
Will look into it, think the message can be made better

hutchig replied:
just a comment, resolving now.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPLoggingHandler.java:
```diff
@@ -0,0 +1,154 @@
+			@Override
+			public void operationComplete(ChannelFuture future) throws Exception {
+				if(future.isSuccess())
+					Tr.event(ctx.channel(), tc, "write (async) finished sucessfully for local: " + ctx.channel().localAddress() + " remote: " + ctx.channel().remoteAddress()+ " wrote: " + bytesToWrite + " bytes");
```
Does this needs a 
if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {

isaacrivriv replied:
Yes it should, nice catch!

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPLoggingHandler.java:
```diff
@@ -0,0 +1,154 @@
+	@Override
+	public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+		if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+			Tr.event(ctx.channel(), tc, "SocketChannel closed, local: " + ctx.channel().localAddress() + " remote: " + ctx.channel().remoteAddress());
```
Inactive is better here than 'closed'

isaacrivriv replied:
This is to match the same logging as CHFW, inactive is fired once the Channel is closed where as close is for the logic where closing is taking place

hutchig replied:
OK

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+                if(config.isInbound()) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Adding new channel group for " + openFuture.channel());
+                    }
```
channel is already cached

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+                	framework.getOutboundConnections().add(channel);
+                }
+
+                // set common channel attrs
```
It is generally better to set up an object's attributes before passing it out of the factory method
or making it visible via a shared data structure
(so here you have added it to the outbound connections or active channels map before setting the
attrs below.)

isaacrivriv replied:
Sure I can make this change

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+        ServerBootstrapExtended bs = new ServerBootstrapExtended();
+        bs.group(framework.getParentGroup(), framework.getChildGroup());
+        bs.channel(NioServerSocketChannel.class);
+        // apply the existing user config to the Netty TCP channel
```
The rest of this could be common with Outbound.

isaacrivriv replied:
Even though the calls are the same, they are different objects one for ServerBootstrap and the other for just Bootstrap both of which we extend and add methods to so can't easily link them both in one for having them in common. Might be good to see if we can reuse some code later on.

isaacrivriv replied:
Created #25412 to follow up

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+        		Tr.debug(tc, TCPMessageConstants.TCP_CHANNEL_STOPPED, channelName, host, String.valueOf(port));
+        } else {
+        	if(TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+        		Tr.debug(tc, "Socket channel closed, local: " + channel.localAddress() + " remote: " + channel.remoteAddress());
```
channelName? and 'stopped' is more correct than 'closed'

isaacrivriv replied:
Made this way to use the same/similar logging as CHFW

hutchig replied:
OK

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/udp/UDPUtils.java:
```diff
@@ -0,0 +1,264 @@
+                channel.attr(ConfigConstants.PORT_KEY).set(inetPort);
+                channel.attr(ConfigConstants.IS_INBOUND_KEY).set(config.isInboundChannel());
+
+                framework.getOutboundConnections().add(channel);
```
here we have another pseudonym:  'Connections' and a 'channel' why not getOutboundChannels as that is what it actually is. Having different names for things just increases the reading age of code.

isaacrivriv replied:
Refer to the previous conversation where I go through the terminology

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+                
+                // Listener to stop channel on close
+                // This should just log that the channel stopped
+                channel.closeFuture().addListener(innerFuture -> logChannelStopped(channel));
```
Comment made in error, a completed future will just run the listener code directly.

hutchig commented on dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/tcp/TCPUtils.java:
```diff
@@ -0,0 +1,312 @@
+                }
+            } else {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc,
```
If an inbound success is logged as a info I think a failure is at least an info/warning rather than debug?

isaacrivriv replied:
The error is still logged as expected a few line after this one if no retries are left. As for the using info/debug and error/debug I believe is to follow the same logging type as CHFW

hutchig replied:
OK

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/.classpath:
```diff
@@ -0,0 +1,8 @@
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="src" path="/io.openliberty.netty.internal"/>
```
I think the class path for the eclipse project should be controlled by bndtools and not checked in
to have specific projects:
<?xml version="1.0" encoding="UTF-8"?>
<classpath>
	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
	<classpathentry kind="output" path="bin"/>
</classpath>

isaacrivriv replied:
I'll look into this

isaacrivriv replied:
I see what you mean, I'll make this change not sure why this showed up here. Probably the IDE adding it there by default

hutchig commented on dev/io.openliberty.netty.internal.tls.impl/src/io/openliberty/netty/internal/tls/impl/NettyTlsProviderImpl.java:
```diff
@@ -0,0 +1,327 @@
+ * String, String, String, Boolean, SSLConnectionLink)}
+ *
+ */
+@Component(configurationPid = "io.openliberty.netty.internal.tls", immediate = true, service = NettyTlsProvider.class, property = { "service.vendor=IBM" })
```
See https://ibm-cloud.slack.com/archives/C31DW78RG/p1633010438267800 re service.vendor

isaacrivriv replied:
I can remove it

hutchig reviewed (dismissed):
All my comments are merely suggestions (have a think about them please),
well done Isaac for such a complex piece of work being done interleaved with all the other plates you are spinning.

isaacrivriv committed:
Fixed review comments

isaacrivriv commented:
#build

isaacrivriv commented:
#build

isaacrivriv commented:
#libby

loriadi reviewed (approved):
I'm approving based on comments from the netty team that the PR looks good.
